File: Heap/heap.ts
let minHeap: number[] = [];
let maxHeap: number[] = [];

// 1 -> 2,3
// 2 -> 4,5
// 3 -> 6,7
// 4 -> 8,9
// 5 -> 10,11

const minHeapify = (insertPosition: number) => {
  let parentIndex = Math.ceil(insertPosition / 2 - 1);
  let swap = minHeap[parentIndex];
  while (minHeap[parentIndex] > minHeap[insertPosition]) {
    swap = minHeap[parentIndex];
    minHeap[parentIndex] = minHeap[insertPosition];
    minHeap[insertPosition] = swap;
    insertPosition = parentIndex;
    parentIndex = Math.ceil(insertPosition / 2 - 1);
  }
};

const maxHeapify = (insertPosition: number) => {
  let parentIndex = Math.ceil(insertPosition / 2 - 1);
  let swap = maxHeap[parentIndex];
  while (maxHeap[parentIndex] < maxHeap[insertPosition]) {
    swap = maxHeap[parentIndex];
    maxHeap[parentIndex] = maxHeap[insertPosition];
    maxHeap[insertPosition] = swap;
    insertPosition = parentIndex;
    parentIndex = Math.ceil(insertPosition / 2 - 1);
  }
};

const addElementMinHeap = (unsortedArray: number[]) => {
  unsortedArray.forEach((ele) => {
    const elementIndex = minHeap.length;
    minHeap.push(ele);
    minHeapify(elementIndex);
  });
};

const addElementMaxHeap = (unsortedArray: number[]) => {
  unsortedArray.forEach((ele) => {
    const elementIndex = maxHeap.length;
    maxHeap.push(ele);
    maxHeapify(elementIndex);
  });
};

const removeElementMinHeap = () => {
  const removeElement = minHeap[0];
  const leafIndex = minHeap.length - 1;
  const leafElement = minHeap[leafIndex];
  minHeap[0] = leafElement;
  minHeap = minHeap.slice(0, leafIndex);
  let parentIndex = 0;
  let leftChildIndex = parentIndex * 2 + 1;
  let rightChildIndex = parentIndex * 2 + 2;
  let swapChildIndex = 0;
  let swap = 0;

  while (
    leafElement > minHeap[leftChildIndex] ||
    leafElement > minHeap[rightChildIndex]
  ) {
    if (
      minHeap[leftChildIndex] < minHeap[rightChildIndex] ||
      !minHeap[rightChildIndex]
    ) {
      swapChildIndex = leftChildIndex;
    } else if (
      minHeap[rightChildIndex] < minHeap[leftChildIndex] ||
      !minHeap[leftChildIndex]
    ) {
      swapChildIndex = rightChildIndex;
    }
    swap = minHeap[swapChildIndex];
    minHeap[swapChildIndex] = leafElement;
    minHeap[parentIndex] = swap;
    parentIndex = swapChildIndex;
    leftChildIndex = parentIndex * 2 + 1;
    rightChildIndex = parentIndex * 2 + 2;
  }

  return removeElement;
};

const removeElementMaxHeap = () => {
  const removeElement = maxHeap[0];
  const leafIndex = maxHeap.length - 1;
  const leafElement = maxHeap[leafIndex];
  maxHeap[0] = leafElement;
  maxHeap = maxHeap.slice(0, leafIndex);
  let parentIndex = 0;
  let leftChildIndex = parentIndex * 2 + 1;
  let rightChildIndex = parentIndex * 2 + 2;
  let swapChildIndex = 0;
  let swap = 0;

  while (
    leafElement < maxHeap[leftChildIndex] ||
    leafElement < maxHeap[rightChildIndex]
  ) {
    if (
      maxHeap[leftChildIndex] > maxHeap[rightChildIndex] ||
      !maxHeap[rightChildIndex]
    ) {
      swapChildIndex = leftChildIndex;
    } else if (
      maxHeap[rightChildIndex] > maxHeap[leftChildIndex] ||
      !maxHeap[leftChildIndex]
    ) {
      swapChildIndex = rightChildIndex;
    }
    swap = maxHeap[swapChildIndex];
    maxHeap[swapChildIndex] = leafElement;
    maxHeap[parentIndex] = swap;
    parentIndex = swapChildIndex;
    leftChildIndex = parentIndex * 2 + 1;
    rightChildIndex = parentIndex * 2 + 2;
  }

  return removeElement;
};

const getMinSortedArray = () => {
  const sortedArray: number[] = [];
  minHeap.forEach((ele) => {
    const minElement = removeElementMinHeap();
    sortedArray.push(minElement);
  });
  console.log("Min Sorted Array", sortedArray);
};

const getMaxSortedArray = () => {
  const sortedArray: number[] = [];
  maxHeap.forEach((ele) => {
    const maxElement = removeElementMaxHeap();
    sortedArray.push(maxElement);
  });
  console.log("Max Sorted Array", sortedArray);
};

const start = () => {
  const unsortedArray = [20, 10, 5, 32, 7, 2, 1, 27, 22];

  addElementMinHeap([...unsortedArray]);
  console.log("MIN HEAP", minHeap);
  getMinSortedArray();

  addElementMaxHeap([...unsortedArray]);
  console.log("MAX HEAP", maxHeap);
  getMaxSortedArray();
};

start();

// 1 -> 2,3
// 2 -> 4,5
// 3 -> 6,7
// 4 -> 8,9
// 5 -> 10,11

//expected Min Heap [ 1, 7,  2, 22, 20, 10, 5, 32, 27 ]

File: Golang/Heap/Array/heap.go
package main

import (
	"fmt"
	"sync"
)

func MinHeapify(elements *[]int, index int) {
	parentIndex := index / 2

	var swap int
	for (*elements)[parentIndex] > (*elements)[index] {
		swap = (*elements)[parentIndex]
		(*elements)[parentIndex] = (*elements)[index]
		(*elements)[index] = swap
		index = parentIndex
		parentIndex = index / 2
	}

}

func MaxHeapify(elements *[]int, index int) {
	parentIndex := index / 2

	var swap int
	for (*elements)[parentIndex] < (*elements)[index] {
		swap = (*elements)[parentIndex]
		(*elements)[parentIndex] = (*elements)[index]
		(*elements)[index] = swap
		index = parentIndex
		parentIndex = index / 2
	}

}

func getChildIndex(parentIndex int) (*int, *int) {
	leftChildIndex := (parentIndex * 2) + 1
	rightChildIndex := (parentIndex * 2) + 2

	return &leftChildIndex, &rightChildIndex
}

func getChildElements(parentIndex int, heap *[]int) (*int, *int) {
	leftChildIndex := (parentIndex * 2) + 1
	rightChildIndex := (parentIndex * 2) + 2

	var leftChildElement *int
	var rightChildElement *int
	heapLen := len(*heap)

	if heapLen > leftChildIndex {
		leftChildElement = &(*heap)[leftChildIndex]
	}
	if heapLen > rightChildIndex {
		rightChildElement = &(*heap)[rightChildIndex]
	}

	return leftChildElement, rightChildElement
}

func removeElementMinHeap(heap *[]int) *int {
	if len(*heap) == 0 {
		return nil
	}
	parentIndex := 0
	removeElement := (*heap)[parentIndex]
	parentElement := (*heap)[len(*heap)-1]
	(*heap)[parentIndex] = parentElement
	*heap = append([]int{}, (*heap)[:len(*heap)-1]...)
	leftChildIndex, rightChildIndex := getChildIndex(parentIndex)
	leftChild, rightChild := getChildElements(parentIndex, heap)

	for leftChild != nil && *leftChild < parentElement || rightChild != nil && *rightChild < parentElement {
		if leftChild != nil && *leftChild < parentElement && rightChild == nil || *leftChild < *rightChild {
			swap := (*heap)[*leftChildIndex]
			(*heap)[*leftChildIndex] = parentElement
			(*heap)[parentIndex] = swap
			parentIndex = *leftChildIndex
			parentElement = (*heap)[parentIndex]
			leftChildIndex, rightChildIndex = getChildIndex(parentIndex)
			leftChild, rightChild = getChildElements(parentIndex, heap)
			continue
		} else if rightChild != nil {
			swap := (*heap)[*rightChildIndex]
			(*heap)[*rightChildIndex] = parentElement
			(*heap)[parentIndex] = swap
			parentIndex = *rightChildIndex
			parentElement = (*heap)[parentIndex]
			leftChildIndex, rightChildIndex = getChildIndex(parentIndex)
			leftChild, rightChild = getChildElements(parentIndex, heap)
			continue
		}
	}
	return &removeElement
}

func removeElementMaxHeap(heap *[]int) *int {
	if len(*heap) == 0 {
		return nil
	}
	parentIndex := 0
	removeElement := (*heap)[parentIndex]
	parentElement := (*heap)[len(*heap)-1]
	(*heap)[parentIndex] = parentElement
	*heap = append([]int{}, (*heap)[:len(*heap)-1]...)
	leftChildIndex, rightChildIndex := getChildIndex(parentIndex)
	leftChild, rightChild := getChildElements(parentIndex, heap)

	for leftChild != nil && *leftChild > parentElement || rightChild != nil && *rightChild > parentElement {
		if leftChild != nil && *leftChild > parentElement && rightChild == nil || *leftChild > *rightChild {
			swap := (*heap)[*leftChildIndex]
			(*heap)[*leftChildIndex] = parentElement
			(*heap)[parentIndex] = swap
			parentIndex = *leftChildIndex
			parentElement = (*heap)[parentIndex]
			leftChildIndex, rightChildIndex = getChildIndex(parentIndex)
			leftChild, rightChild = getChildElements(parentIndex, heap)
			continue
		} else if rightChild != nil {
			swap := (*heap)[*rightChildIndex]
			(*heap)[*rightChildIndex] = parentElement
			(*heap)[parentIndex] = swap
			parentIndex = *rightChildIndex
			parentElement = (*heap)[parentIndex]
			leftChildIndex, rightChildIndex = getChildIndex(parentIndex)
			leftChild, rightChild = getChildElements(parentIndex, heap)
			continue
		}
	}
	return &removeElement
}

func getMinSortedArray(elements *[]int, wg *sync.WaitGroup) *[]int {
	heap := make([]int, 0)
	sortedArray := make([]int, 0)

	for i, v := range *elements {
		heap = append(heap, v)
		MinHeapify(&heap, i)
	}

	fmt.Println("Min heap --->", heap)

	for range *elements {
		removedElement := removeElementMinHeap(&heap)
		sortedArray = append(sortedArray, (*removedElement))
	}

	fmt.Println("Sorted Element --->", sortedArray)
	defer wg.Done()
	return &sortedArray

}

func getMaxSortedArray(elements *[]int, wg *sync.WaitGroup) *[]int {
	heap := make([]int, 0)
	sortedArray := make([]int, 0)

	for i, v := range *elements {
		heap = append(heap, v)
		MaxHeapify(&heap, i)
	}

	fmt.Println("Max heap --->", heap)

	for range *elements {
		removedElement := removeElementMaxHeap(&heap)
		sortedArray = append(sortedArray, (*removedElement))
	}

	fmt.Println("Sorted Element --->", sortedArray)
	defer wg.Done()
	return &sortedArray
}

func Sort(elements *[]int) {
	wg := &sync.WaitGroup{}

	wg.Add(2)

	go getMinSortedArray(elements, wg)
	go getMaxSortedArray(elements, wg)
	wg.Wait()
}

func main() {
	unsortedElements := []int{3, 2, 1, 6, 8, 4, 5, 2, 1, 6}
	Sort(&unsortedElements)
}
